File: src/main/java/com/server/ordering/service/CouponService.java
package com.server.ordering.service;

import com.server.ordering.domain.Coupon;
import com.server.ordering.domain.MyCoupon;
import com.server.ordering.domain.dto.request.CouponDto;
import com.server.ordering.domain.member.Customer;
import com.server.ordering.repository.CouponRepository;
import com.server.ordering.repository.CustomerRepository;
import com.server.ordering.repository.MyCouponRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.NoResultException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
@Slf4j
public class CouponService {

    private final CouponRepository couponRepository;
    private final MyCouponRepository myCouponRepository;
    private final CustomerRepository customerRepository;

    @Transactional
    public void saveCoupon(CouponDto couponDto) {
        Coupon coupon = new Coupon(couponDto.getSerialNumber(), couponDto.getValue());
        couponRepository.save(coupon);
    }

    public Coupon getCoupon(String id) {
        Coupon coupon = couponRepository.find(id);
        if (coupon == null) {
            throw new NoResultException();
        }
        return coupon;
    }

    public Boolean haveThisCoupon(String serialNumber, Long customerId) {
        try {
            myCouponRepository.findOneWithCustomer(serialNumber, customerId);
            return true;
        } catch (NoResultException e) {
            return false;
        }
    }

    @Transactional
    public void useMyCoupon(Long couponId) {
        myCouponRepository.remove(couponId);
    }

    @Transactional
    public void saveMyCoupon(Coupon coupon, Long customerId) {
        Customer customer = customerRepository.findOne(customerId);
        MyCoupon myCoupon = new MyCoupon(coupon.getSerialNumber(), coupon.getValue(), customer);
        myCouponRepository.save(myCoupon);
    }
}

File: src/main/kotlin/com/byeoru/ordering_server/service/CouponService.kt
package com.byeoru.ordering_server.service

import com.byeoru.ordering_server.domain.Coupon
import com.byeoru.ordering_server.domain.MyCoupon
import com.byeoru.ordering_server.domain.dto.request.CouponDto
import com.byeoru.ordering_server.repository.CouponRepository
import com.byeoru.ordering_server.repository.CustomerRepository
import com.byeoru.ordering_server.repository.MyCouponRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.persistence.NoResultException

@Service
@Transactional(readOnly = true)
class CouponService(private val couponRepository: CouponRepository,
                    private val myCouponRepository: MyCouponRepository,
                    private val customerRepository: CustomerRepository) {

    @Transactional
    fun saveCoupon(couponDto: CouponDto) {
        val coupon = Coupon(couponDto.serialNumber, couponDto.value)
        couponRepository.save(coupon)
    }

    fun getCoupon(id: String): Coupon {
        return couponRepository.find(id)
    }

    fun haveThisCoupon(serialNumber: String?, customerId: Long?): Boolean {
        return try {
            myCouponRepository.findOneWithCustomer(serialNumber, customerId)
            true
        } catch (e: NoResultException) {
            false
        }
    }

    @Transactional
    fun useMyCoupon(couponId: Long) = myCouponRepository.remove(couponId)

    @Transactional
    fun saveMyCoupon(coupon: Coupon, customerId: Long) {
        val customer = customerRepository.findOne(customerId)
        val myCoupon = MyCoupon(coupon.serialNumber, coupon.value, customer)
        myCouponRepository.save(myCoupon)
    }

    companion object {
        private val log = LoggerFactory.getLogger(CouponService::class.java)
    }
}